core/node/groups.go

```diff
@@ -2,105 +2,284 @@ package node
 	offroute "github.com/ipfs/go-ipfs-routing/offline"
 	"github.com/ipfs/go-path/resolver"
+	uio "github.com/ipfs/go-unixfs/io"
 	"go.uber.org/fx"
 )
 
 var BaseLibP2P = fx.Options(
-	fx.Provide(libp2p.AddrFilters),
-	fx.Provide(libp2p.BandwidthCounter),
 	fx.Provide(libp2p.PNet),
-	fx.Provide(libp2p.AddrsFactory),
 	fx.Provide(libp2p.ConnectionManager),
-	fx.Provide(libp2p.NatPortMap),
-	fx.Provide(libp2p.Relay),
-	fx.Provide(libp2p.AutoRealy),
 	fx.Provide(libp2p.DefaultTransports),
-	fx.Provide(libp2p.QUIC),
 
 	fx.Provide(libp2p.Host),
 
 	fx.Provide(libp2p.DiscoveryHandler),
 
-	fx.Invoke(libp2p.AutoNATService),
 	fx.Invoke(libp2p.PNetChecker),
-	fx.Invoke(libp2p.StartListening),
-	fx.Invoke(libp2p.SetupDiscovery),
 )
 
-func LibP2P(cfg *BuildCfg) fx.Option {
+func LibP2P(bcfg *BuildCfg, cfg *config.Config) fx.Option {
+	// parse ConnMgr config
+
+	grace := config.DefaultConnMgrGracePeriod
+	low := config.DefaultConnMgrHighWater
+	high := config.DefaultConnMgrHighWater
+
+	connmgr := fx.Options()
+
+	if cfg.Swarm.ConnMgr.Type != "none" {
+		switch cfg.Swarm.ConnMgr.Type {
+		case "":
+			// 'default' value is the basic connection manager
+			break
+		case "basic":
+			var err error
+			grace, err = time.ParseDuration(cfg.Swarm.ConnMgr.GracePeriod)
+			if err != nil {
+				return fx.Error(fmt.Errorf("parsing Swarm.ConnMgr.GracePeriod: %s", err))
+			}
+
+			low = cfg.Swarm.ConnMgr.LowWater
+			high = cfg.Swarm.ConnMgr.HighWater
+		default:
+			return fx.Error(fmt.Errorf("unrecognized ConnMgr.Type: %q", cfg.Swarm.ConnMgr.Type))
+		}
+
+		connmgr = fx.Provide(libp2p.ConnectionManager(low, high, grace))
+	}
+
+	// parse PubSub config
+
+	ps := fx.Options()
+	if bcfg.getOpt("pubsub") || bcfg.getOpt("ipnsps") {
+		var pubsubOptions []pubsub.Option
+		if cfg.Pubsub.DisableSigning {
+			pubsubOptions = append(pubsubOptions, pubsub.WithMessageSigning(false))
+		}
+
+		if cfg.Pubsub.StrictSignatureVerification {
+			pubsubOptions = append(pubsubOptions, pubsub.WithStrictSignatureVerification(true))
+		}
+
+		switch cfg.Pubsub.Router {
+		case "":
+			fallthrough
+		case "floodsub":
+			ps = fx.Provide(libp2p.FloodSub(pubsubOptions...))
+		case "gossipsub":
+			ps = fx.Provide(libp2p.GossipSub(pubsubOptions...))
+		default:
+			return fx.Error(fmt.Errorf("unknown pubsub router %s", cfg.Pubsub.Router))
+		}
+	}
+
+	// Gather all the options
+
 	opts := fx.Options(
 		BaseLibP2P,
 
-		fx.Provide(libp2p.Security(!cfg.DisableEncryptedConnections)),
-		maybeProvide(libp2p.Pubsub, cfg.getOpt("pubsub") || cfg.getOpt("ipnsps")),
+		fx.Provide(libp2p.AddrFilters(cfg.Swarm.AddrFilters)),
+		fx.Invoke(libp2p.SetupDiscovery(cfg.Discovery.MDNS.Enabled, cfg.Discovery.MDNS.Interval)),
+		fx.Provide(libp2p.AddrsFactory(cfg.Addresses.Announce, cfg.Addresses.NoAnnounce)),
+		fx.Provide(libp2p.SmuxTransport(bcfg.getOpt("mplex"))),
+		fx.Provide(libp2p.Relay(cfg.Swarm.DisableRelay, cfg.Swarm.EnableRelayHop)),
+		fx.Invoke(libp2p.StartListening(cfg.Addresses.Swarm)),
```

Comment:
We can handle this later but I wonder if this should go elsewhere; somewhere higher up the stack.